### PR TITLE
[ui, sso] warn user about pending token expiry

### LINKED ISSUE #15061: [ui, sso] Handle token TTL ending gracefully

### PR DESCRIPTION
**Pull request by philrenaud**
Throw a global notification when an active token is within 10 minutes of expiring, and prompt the user to re-authenticate.

These notifications pair with the tokens page changes from #15073 to resolve #15061; these notifications will be seen anyplace in the UI.

image
image
image

### TODO
- [ ] Notification tests
- [ ] Mirage soon-expiring token

### CODE REVIEW

File: ui/app/services/token.js [hunk 1]
```diff
@@ -115,7 +120,67 @@ export default class TokenService extends Service {
...
+              message: `Your token will need to be re-authenticated`,
+            });
+          }
+          this.monitorTokenTime.cancelAll(); // Stop updating time left after expiration
+        }
+      }
+      yield timeout(1000);
+    }
+  })
+  monitorTokenTime;
```
Comment (philrenaud) [inline ui/app/services/token.js:183]: This is using a syntax common to Nomad's front-end, inherited from [Ember Concurrency](http://ember-concurrency.com/docs/introduction/). There were a few attempts to handle this using native JS async/await but, despite the large block size, this ended up feeling more streamlined.

File: ui/app/styles/core/notifications.scss [hunk 2]
```diff
@@ -1,3 +1,5 @@
+$bonusRightPadding: 20px;
```
Comment (philrenaud) [inline ui/app/styles/core/notifications.scss:1]: variablizing the px-specific space we hold open for our close button

File: ui/app/templates/application.hbs [hunk 3]
```diff
@@ -8,13 +8,16 @@
...
+    <span class="close-button" role="button" {{on "click" (queue (action close) (action flash.customCloseAction))}}>&times;</span>
     {{#if flash.title}}
     <h3>{{flash.title}}</h3>
     {{/if}}
     {{#if flash.message}}
       <p>{{flash.message}}</p>
     {{/if}}
+    {{#if flash.customAction}}
+      <button type="button" class="button custom-action-button" {{on "click" (action flash.customAction.action)}}>{{flash.customAction.label}}</button>
+    {{/if}}
```
Comment (philrenaud) [inline ui/app/templates/application.hbs:20]: Added a couple spots in our notifications for custom actions; both optional. One on close and another that appears as a button in the notification itself.

File: ui/app/templates/settings/tokens.hbs [hunk 4]
```diff
@@ -41,48 +41,49 @@
...
-          <div class="columns">
-            <div class="column">
-              <h3 class="title is-4">Token Failed to Authenticate</h3>
-              <p>The token secret you have provided does not match an existing token.</p>
+        {{#if this.tokenIsInvalid}}
+          <div data-test-token-error class="notification is-danger">
+            <div class="columns">
+              <div class="column">
+                <h3 class="title is-4">Token Failed to Authenticate</h3>
+                <p>The token secret you have provided does not match an existing token, or has expired.</p>
```
Comment (philrenaud) [inline ui/app/templates/settings/tokens.hbs:81]: Review this file with "Hide Whitespace" enabled for the sake of sanity; this block was just moved into an `...unless` of the token.isExpired block from before.

**Review by jrasell — APPROVED** [review]
LGTM.

would be worth adding a changelog entry for this.

**Pull request merged**